waveorder/stokes.py

```diff
@@ -0,0 +1,344 @@
+    s012_CPL_after_AR (see tests for examples), and matches the orientation
+    range convention used in comp-micro: 0 <= ori < pi.
+
+    Parameters
+    ----------
+    s1, s2: NDArray, identical shapes
+        Stokes parameters
+
+    Returns
+    -------
+    NDArray
+        Slow-axis orientation with 0 <= ori < pi.
+    """
+    return (np.arctan2(s1, -s2) % (2 * np.pi)) / 2
+
+
+def inverse_s0123_CPL_after_ADR(s0, s1, s2, s3):
+    """
+    Inverse of s0123_CPL_after_ADR.
+
+    Given the Stokes parameters of circularly polarized light (CPL) that has
+    passed through an attenuating depolarizing retarder (ADR), this function
+    returns the parameters of the ADR, specifically its retardance (ret),
+    slow-axis orientation (ori), transmittance (tra), and depolarization (dop).
+
+    Note: this function is commonly used in QLIPP-type reconstructions. After
+    converting raw intensities into Stokes parameters by applying A_inv, the
+    Stokes parameters are uses to estimate the parameters of the sample ADR in
+    a single step with this function.
+
+    Parameters
+    ----------
+    s0, s1, s2, s3: NDArray, identical shapes
+        Stokes parameters
+
+    Returns
+    ----------
+    ret, ori, tra, dop: NDArray
+        ret: retardance of ADR, 2*pi periodic
+        ori: slow-axis orientation of ADR, 2*pi periodic
+        tra: transmittance of ADR, 0 <= tra <= 1
+        dop: depolarization of ADR, 0 <= dop <= 1
+    """
+    len_pol = (s1**2 + s2**2 + s3**2) ** 0.5
+    ret = np.arcsin(((s1**2 + s2**2) ** 0.5) / len_pol)
+    ori = _s12_to_ori(s1, s2)
+    tra = s0
+    dop = len_pol / s0
+    return ret, ori, tra, dop
+
+
+def inverse_s012_CPL_after_AR(s0, s1, s2):
+    """
+    Inverse of s012_CPL_after_ADR.
+
+    Given the Stokes parameters of circularly polarized light (CPL) that has
+    passed through an attenuating retarder (AR), this function returns the
+    parameters of the ADR, specifically its retardance (ret), slow-axis
+    orientation (ori), and transmittance (tra).
+
+    Parameters
+    ----------
+    s0, s1, s2: NDArray, identical shapes
+        First three Stokes parameters
+
+    Returns
+    ----------
+    ret, ori, tra: NDArray, identical shapes
+        ret: retardance of ADR, 2*pi periodic
+        ori: slow-axis orientation of ADR, 2*pi periodic
+        tra: transmittance of ADR, 0 <= tra <= 1
+    """
+    ret = np.arcsin(((s1**2 + s2**2) ** 0.5) / s0)
+    ori = _s12_to_ori(s1, s2)
+    tra = s0
+    return ret, ori, tra
+
+
+def AR_mueller_from_CPL_projection(s0, s1, s2, s3):
+    """
+    Given the Stokes parameters of circularly polarized light (CPL) that has
+    passed through an attenuating retarder (AR), this function returns the
+    complete Mueller matrix of the AR.
+
+    Parameters
+    ----------
+    s0, s1, s2, s3 : NDArray, identical shapes
+        Stokes parameters
+
+    Returns
+    -------
+    NDArray, float, M.shape = (4, 4,) + s0.shape
```

Comment:
Can $s_{i}$ be 3D for this function? If not I suggest changing this to:

```
Returns
-------
NDArray[np.floating]
    Mueller matrix, shape is (4, 4, Y, X), where (Y, X) is the shape of `s0`.
```

Also note that if Python `float` (usually `float64`) is not strictly needed, [a more generic type](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.floating) can be used.

Reply:
s_i can be 1D, 2D, or 3D, but most commonly 1D or 2D. I think I'll leave it as is. 

Removed `float` since I don't think it's necessary to specify.

Reply:
I'm going to reinstate `float` in the docs since it's convenient to have the option to give an ndarray or a float. I ran into this use case myself during testing.

Reply:
Sorry I misunderstood the typing again... didn't realize it was a union, not a member type.